File: 1602-find-nearest-right-node-in-binary-tree/1602-find-nearest-right-node-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode findNearestRightNode(TreeNode root, TreeNode u) {
        LinkedList<TreeNode> que = new LinkedList<>();
        que.addLast(root);
        TreeNode ans = null;
        //Level Order Traversal
        while(que.size()!=0) {
            int size = que.size();
            //Set up the next level in que
            for(int i=0;i<size;i++) {
                TreeNode removeN = que.removeFirst();
                
                //add left and right nodes if exist
                if(removeN.left != null) {
                    que.addLast(removeN.left);
                }
                
                if(removeN.right !=null) {
                    que.addLast(removeN.right);
                }
                //Check if current removeN is our U node
                if(removeN.val == u.val && i!=size-1) {
                    ans = que.peekFirst();
                }
            }
        } 
        return ans;
    }
}

File: 1602-find-nearest-right-node-in-binary-tree/1602-find-nearest-right-node-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* findNearestRightNode(TreeNode* root, TreeNode* u) {
        queue<TreeNode*> que;
        que.push(root);
        TreeNode* ans = nullptr;
        while(que.size()!=0) {
            int size = que.size();
            for(int i=0;i<size;i++) {
                TreeNode* removeN = que.front(); que.pop();
                
                if(removeN->left) que.push(removeN->left);
                if(removeN->right) que.push(removeN->right);
                
                if(removeN->val == u->val && i != size-1) ans = que.front();
            }
        }
        return ans;
    }
};